Step4 리뷰 요청 드립니다.

itdar opened the pull request:
안녕하세요 선흠님! 4단계 리뷰요청 드립니다. 간단하게 생각을 했는데 까다롭고 어려웠습니다 ㅠㅠ

 스프링의 MethodParameter 를 쓸까 하다가.. 필요한 것만 가볍게 하려고 별도로 클래스 만들었습니다.
제가 이해한 바로는 MethodParameter 클래스가, 메서드에 있는 하나하나의 파라미터를 래핑한 헬퍼클래스로 이해하고 만들었습니다! 제대로 이해한게 맞을런지요..ㅎㅎ;;
마찬가지로 스프링의 HandlerMethodArgumentResolver 를 따라해서 인터페이스를 만들었고, 등록은 config 가 아니라 Composite 클래스에 등록하도록 구현하였습니다!

 구현 시에는 HandlerMethodArgumentResolverTest 에서 테스트를 각각 만들어서 확인하며 구현했고,
최종적으롤 기존 UserController 를 대체할 수 있는만큼 대체하였습니다.
 이 때, request.setAttribute() 때문에 서블릿요청 파라미터를 남겨둔게 좀 있습니다. 응답도 아니고 요청에 컨트롤러에서 왜 set 을 해주는지 모르겠습니다..

 리뷰를 받고 좀 정리를 다시하며 리팩토링 진행해보고 싶어서 리뷰를 먼저 요청드립니다!
감사합니다!!! 남은 연휴 잘 보내세요!!

itdar committed:
docs: 4단계 요구사항 정리

itdar committed:
feat: 4단계 Controller 메소드 인자 매핑 구현

itdar committed:
chore: 기존의 UserController 에서 추가된 MethodArgumentHandlerResolver 적용

cogito90 commented on src/main/java/core/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -18,7 +18,7 @@ public class AnnotationHandlerMapping {
     private Object[] basePackage;
 
-    private Map<HandlerKey, HandlerAdapter> handlerExecutions = Maps.newHashMap();
+    private Map<String, HandlerAdapter> handlerExecutions = Maps.newHashMap();
```
👍

cogito90 commented on src/main/java/core/mvc/tobe/HandlerAdapter.java:
```diff
@@ -1,21 +1,29 @@
 
     public ModelAndView handle(HttpServletRequest request, HttpServletResponse response) throws Exception {
-        return (ModelAndView) method.invoke(declaredObject, request, response);
+        Object[] values = resolveParameters(method, request, response);
```
이건 취향일 것 같은데요! 전 개인적으로 굉장히 common한 메서드들이 아니면 static import를 안 하는 편입니다! ( `Objects.isNull`, `Collectors.toList` 등등... ) 이름이 생소하면 얘가 private인지 static 메서드인지 잘 구분이 안가서요! 형진님 의견은 어떠신가요 ? :)

itdar replied:
엇 저도 그냥 핸들러어댑터에서 생성을 해서 필드로 넣어둘까 했었습니다!
그런데 현재 구조상 핸들러어댑터가 컨트롤러의 메서드마다 생성되게 되어있어서
HandlerMethodArgumentResolverComposite 에서 상태없이 메서드 실행만 하는데 여러개 생성 할 필요가 없을 것 같아서..
enum 으로 싱글턴 객체를 만들까, static 메서드로 만들까 하다가 편리함을 따라가벼렀네요....ㅎㅎ;;

cogito90 replied:
앗, 의도가 잘못 전달된 것 같아서 추가로 남겨봅니다!
말씀드렸던 건, static import에 대한 얘기였어요!
static하게 접근하는 건 둘째치고, 직접 정의해서 만든 Util 클래스들은 static import를 하게 되면 다른사람이 보기엔 static하게 접근한건지, private 메서드를 참조한건지 좀 헷갈릴 수가 있지 않을까 하는 의견이었습니다 :)

itdar replied:
아아 그러네요 ㅎㅎ;; 클래스명까지 넣는게 확실해서 더 좋을 것 같습니다!!

cogito90 commented on src/main/java/next/controller/IndexController.java:
```diff
@@ -13,10 +13,24 @@
         return new ModelAndView(jspView);
     }
 
+    @RequestMapping(value = "/index.html", method = RequestMethod.GET)
```
요거는 약간 RequestMapping으로 핸들링 되어야 하는 부분이 아닌 것 같은데요 ?

itdar replied:
엇 맞아요... 흑흑 제거하고 아래랑 같이 메서드 분리해서 하나로 처리해보겠습니다 ㅠ.ㅠ

cogito90 replied:
요것도 약간 방향이 다르게 전달된 것 같아서..!
url이 index`.html` 로 되어있어서, static resource 를 반환하는 컨트롤러 쪽으로 넘어가야 하지 않나 싶어서요!

itdar replied:
아아 정적리소스로!! 옙옙

cogito90 commented on src/main/java/next/controller/IndexController.java:
```diff
@@ -13,10 +13,24 @@
 
+    @RequestMapping(value = "/index.html", method = RequestMethod.GET)
+    public ModelAndView indexHtml(HttpServletRequest req, HttpServletResponse resp) {
+        req.setAttribute("users", DataBase.findAll());
```
그리고 밑에 내용이 동일해서 url은 다르더라도 같은 동작을 하는거라면 메서드 세퍼레이션을 통해 하나의 메서드로 관리할 수 있을 것 같아요!

cogito90 commented on src/main/java/core/mvc/tobe/MethodParameter.java:
```diff
@@ -0,0 +1,70 @@
+
+public class MethodParameter {
+
+    private ParameterNameDiscoverer nameDiscoverer = new LocalVariableTableParameterNameDiscoverer();
```
요거 static으로 선언해서 객체 생성시 부하를 조금 줄여볼 수 있지 않을까요 ? 내부적으로 동시접근도 제어 하고 있는 것 같아요!

itdar replied:
오 옙! 혹시 내부적으로 동시접근 제어하는 부분은 어디서 확인할 수 있을까요? 
내부가 복잡해서 잘 못찾겠네요..ㅎㅎ;;

cogito90 replied:
`LocalVariableTableParameterNameDiscoverer` 내부에서 ConcurrentMap 사용하고 있더라구요!

itdar replied:
오 옙 감사합니다!

cogito90 commented on src/main/java/core/mvc/tobe/MethodParameter.java:
```diff
@@ -0,0 +1,70 @@
+    private int parameterIndex;
+    private Annotation[] annotations;
+
+    public MethodParameter(Method method, int parameterIndex) {
```
Test말곤 사용처가 없는걸까요 ?

itdar replied:
옙 이 생성자 자체는 최초 구현 시에 테스트케이스 기반으로 어노테이션 없는 메서드 생성 시 만들었던 것이네요..
체인으로 사용하여 그냥 남겨두었습니다!!

cogito90 commented on src/main/java/core/mvc/tobe/MethodParameter.java:
```diff
@@ -0,0 +1,70 @@
+    private ParameterNameDiscoverer nameDiscoverer = new LocalVariableTableParameterNameDiscoverer();
+
+    private Method method;
+    private int parameterIndex;
```
혹시 index가 이 클래스 내부에 필요한 이유가 있을까요 ? 외부에서의 사용처가 없는 것 같아서요!

itdar replied:
아 스프링의 MethodParameter 클래스의 필드를 참고했었는데요
현재 index 사용하는 곳은 method 에서 parameter 의 타입, 이름 을 가져올 때 index 를 사용하고 있습니다!!
( getParameterType(), getParameterName() )

cogito90 commented on src/main/java/core/mvc/tobe/argumentresolver/DefaultMethodArgumentResolver.java:
```diff
@@ -0,0 +1,51 @@
+            return true;
+        }
+
+        if (!methodParameter.hasAnnotation()
```
hasAnnotation을 볼 필요가 있을까요 ? 이름이 Default~~ 인 걸 봐서, Resolver 지정이 안 되는 annotation에 대해서 fallback으로 동작할 것 같은데 그러면 annotation이 붙어 있더라도 무시하면 되니까 필요가 없어 보여서요!

itdar replied:
엇 그러네요 ㅎㅎ;;
마지막에 path variable 어노테이션 관련 내용을 수정하면서 int_long 관련 버그도 생겼던 것 같은데..
같이 수정해보겠습니다!!

cogito90 commented on src/main/java/core/mvc/tobe/argumentresolver/HandlerMethodArgumentResolverComposite.java:
```diff
@@ -0,0 +1,41 @@
+
+            for (HandlerMethodArgumentResolver resolver : RESOLVER_LIST) {
+                if (resolver.supportsParameter(methodParameter)) {
+                    values[i] = resolver.resolveArgument(methodParameter, request, null);
```
response는 null이 맞나요 ?

itdar replied:
엇 response 넘겨주는 것으로 수정했습니다!!

cogito90 commented on src/main/java/core/mvc/tobe/argumentresolver/DefaultMethodArgumentResolver.java:
```diff
@@ -0,0 +1,51 @@
+            return Long.parseLong(value);
+        }
+
+        return value;
```
지금은 double, float, boolean 등은 문자열로 리턴되겠군요 ?

itdar replied:
게으른 것을 반성하고 추가해보았읍니다..ㅎㅎ;;

cogito90 replied:
ㅎㅎ 현재 상황에만 맞춰서 있는것도 괜찮긴 한데 fallback으로 무조건 문자열이 내려가 버리니 모르고 쓰는 경우엔 예기치 못하게 파라미터 바인딩이 안 될 수 있을 것 같아서 말씀 드려 봤어요! 저희는 지금 라이브러리를 제공하는 입장에서 미션을 진행하는거라 ..!

cogito90 commented on src/main/java/core/mvc/tobe/argumentresolver/HandlerMethodArgumentResolverComposite.java:
```diff
@@ -0,0 +1,41 @@
+                    values[i] = resolver.resolveArgument(methodParameter, request, null);
+                    break;
+                }
+                values[i] = request.getParameter(methodParameter.getParameterName());
```
default resolver가 있어서, fallback은 거기서 처리하는게 맞지 싶은데요!

그렇게 되면 stream으로 더 깔끔하게 표현 할 수 있을 것 같아요!

itdar replied:
옙 감사합니다!! 위에 hasAnnotation 처리와 같이 fallback 처리해서 정리해보겠습니다!!

itdar replied:
오 위에 리뷰내용과 이것 까지해서 적용하면서 수정하니 굉장히 깔끔해질 수 있었네요..!!
감사합니다!! ㅎㅎ

shared-moon reviewed (changes requested):
형진님 안녕하세요!
4단계 미션 잘 진행 해 주셨네요 :)
마지막 미션이기도 하고, 형진님 개인적으로도 아직 완성이 덜 된 느낌이라 한번만 더 살펴보면 좋을 것 같아요!
코멘트를 몇 개 남겨 두었는데, 확인 후 리뷰 다시 요청 부탁드릴게요!
화이팅하세요 🔥

itdar commented:
> 형진님 안녕하세요! 4단계 미션 잘 진행 해 주셨네요 :) 마지막 미션이기도 하고, 형진님 개인적으로도 아직 완성이 덜 된 느낌이라 한번만 더 살펴보면 좋을 것 같아요! 코멘트를 몇 개 남겨 두었는데, 확인 후 리뷰 다시 요청 부탁드릴게요! 화이팅하세요 🔥

 UserController 의 메서드들에서 req.getSession() 해서 사용하는 것들은 처리방법을 생각해봤는데..
당장은 전역 context 에서 request 를 뽑아올 방법이 없을 것 같아서
HttpSession 자체를 Resolver 추가해서 받아도 될 것 같고
아니면 @LoginUser 와 같은 커스텀 어노테이션으로 Resolver 를 추가해서 인자로 받을 수도 있을 것 같고..
그렇게 생각을 해보았는데 어떨까요?

itdar committed:
refactor: 선흠님 리뷰사항 수정

- HandlerMethodArgumentResolver 찾는 로직 정리
- static 처리 (LocalVariableTableParameterNameDiscoverer)
- index.html 컨트롤러에서 처리 제거

shared-moon commented:
안녕하세요!
Session까지 고민을 해주셨군요 ..!
말씀하신대로 HttpRequest나 HttpResponse 처리한 것과 같이 HttpSession을 추가로 정의 해 주는게 현재는 가장 간편한 방법일 것 같습니다! :)

cogito90 commented on src/main/java/core/mvc/tobe/argumentresolver/HandlerMethodArgumentResolverComposite.java:
```diff
@@ -25,14 +25,11 @@ public static Object[] resolveParameters(Method method, HttpServletRequest reque
-            }
-
+            values[i] = RESOLVER_LIST.stream()
+                .filter(r -> r.supportsParameter(methodParameter))
```
👍

cogito90 commented on src/main/java/core/mvc/tobe/argumentresolver/DefaultMethodArgumentResolver.java:
```diff
@@ -27,25 +27,20 @@ public Object resolveArgument(MethodParameter methodParameter, HttpServletReques
             return Long.parseLong(value);
         }
 
+        if (parameterType.equals(double.class)) {
```
👍

cogito90 commented on src/main/java/core/mvc/tobe/MethodParameter.java:
```diff
@@ -56,11 +56,7 @@ public Class<?> getParameterType() {
-
-    public boolean hasAnnotation() {
-        return method.getDeclaredAnnotations().length > 0;
+        return NAME_DISCOVERER.getParameterNames(this.method)[parameterIndex];
```
👍

cogito90 commented on src/main/java/core/mvc/tobe/argumentresolver/HandlerMethodArgumentResolverComposite.java:
```diff
@@ -11,8 +11,8 @@
 public class HandlerMethodArgumentResolverComposite {
 
     private static final List<HandlerMethodArgumentResolver> RESOLVER_LIST = new ArrayList<>();
+    private static final DefaultMethodArgumentResolver DEFAULT_RESOLVER = new DefaultMethodArgumentResolver();
```
👍

cogito90 commented on src/main/java/next/controller/IndexController.java:
```diff
@@ -14,20 +14,15 @@ public class IndexController {
-        req.setAttribute("users", DataBase.findAll());
-        JspView jspView = new JspView("home.jsp");
-        return new ModelAndView(jspView);
+        return goToHomeJsp(req);
```
👍

shared-moon reviewed (approved):
형진님 안녕하세요!
피드백 반영 잘 해 주셨네요!
저번 코멘트에 따로 답글을 남겨 두었는데, 다음 미션 진행 전 확인 부탁드려요!
고생 많으셨습니다 :)

itdar commented:
> 형진님 안녕하세요! 피드백 반영 잘 해 주셨네요! 저번 코멘트에 따로 답글을 남겨 두었는데, 다음 미션 진행 전 확인 부탁드려요! 고생 많으셨습니다 :)

감사합니다 선흠님
리뷰해주시느라 고생하셨습니다.
감사합니다!!!!
